diff --git a/bfd/Makefile.am b/bfd/Makefile.am
index 4987ac9ccc..9cba52af2c 100644
--- a/bfd/Makefile.am
+++ b/bfd/Makefile.am
@@ -200,6 +200,7 @@ ALL_MACHINES_CFILES = \
 	cpu-h8300.c \
 	cpu-hppa.c \
 	cpu-i386.c \
+	cpu-i51.c \
 	cpu-ia64.c \
 	cpu-iamcu.c \
 	cpu-ip2k.c \
@@ -448,6 +449,7 @@ BFD32_BACKENDS_CFILES = \
 	elf32-h8300.c \
 	elf32-hppa.c \
 	elf32-i386.c \
+	elf32-i51.c \
 	elf32-ip2k.c \
 	elf32-iq2000.c \
 	elf32-lm32.c \
diff --git a/bfd/archures.c b/bfd/archures.c
index fd22c94d6d..4efb255af4 100644
--- a/bfd/archures.c
+++ b/bfd/archures.c
@@ -210,6 +210,7 @@ DESCRIPTION
 .#define bfd_mach_iamcu			(1 << 8)
 .#define bfd_mach_i386_iamcu		(bfd_mach_i386_i386 | bfd_mach_iamcu)
 .#define bfd_mach_i386_iamcu_intel_syntax (bfd_mach_i386_iamcu | bfd_mach_i386_intel_syntax)
+.  bfd_arch_i51,       {* MCS-51/8051 microcontrollers.  *}
 .  bfd_arch_romp,      {* IBM ROMP PC/RT.  *}
 .  bfd_arch_convex,    {* Convex.  *}
 .  bfd_arch_m98k,      {* Motorola 98xxx.  *}
@@ -649,6 +650,7 @@ extern const bfd_arch_info_type bfd_h8300_arch;
 extern const bfd_arch_info_type bfd_hppa_arch;
 extern const bfd_arch_info_type bfd_i386_arch;
 extern const bfd_arch_info_type bfd_iamcu_arch;
+extern const bfd_arch_info_type bfd_i51_arch;
 extern const bfd_arch_info_type bfd_ia64_arch;
 extern const bfd_arch_info_type bfd_ip2k_arch;
 extern const bfd_arch_info_type bfd_iq2000_arch;
@@ -737,6 +739,7 @@ static const bfd_arch_info_type * const bfd_archures_list[] =
     &bfd_hppa_arch,
     &bfd_i386_arch,
     &bfd_iamcu_arch,
+    &bfd_i51_arch,
     &bfd_ia64_arch,
     &bfd_ip2k_arch,
     &bfd_iq2000_arch,
diff --git a/bfd/bfd-in2.h b/bfd/bfd-in2.h
index bcff44a0f4..c708ee7568 100644
--- a/bfd/bfd-in2.h
+++ b/bfd/bfd-in2.h
@@ -1431,6 +1431,7 @@ enum bfd_architecture
 #define bfd_mach_iamcu                 (1 << 8)
 #define bfd_mach_i386_iamcu            (bfd_mach_i386_i386 | bfd_mach_iamcu)
 #define bfd_mach_i386_iamcu_intel_syntax (bfd_mach_i386_iamcu | bfd_mach_i386_intel_syntax)
+  bfd_arch_i51,       /* MCS-51/8051 microcontrollers.  */
   bfd_arch_romp,      /* IBM ROMP PC/RT.  */
   bfd_arch_convex,    /* Convex.  */
   bfd_arch_m98k,      /* Motorola 98xxx.  */
@@ -5265,6 +5266,39 @@ enum bfd_reloc_code_real
      as a global register.  It does not modify contents.  */
   BFD_RELOC_MMIX_LOCAL,
 
+  /* This is a 1 bit relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_R1,
+
+  /* This is a 3 bit relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_R3,
+
+  /* This is a 7 bit PC relative relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_7_PCREL,
+
+  /* This is an 11 bit relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_11,
+
+  /* This is an 8 bit relocation for bit addressing in the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_8_BIT,
+
+  /* This is an 8 bit relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_8,
+
+  /* This is an 8 bit low byte relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_8_LOW,
+
+  /* This is an 8 bit high byte relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_8_HIGH,
+
+  /* This is a 16 bit relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_16,
+
+  /* This is an 8 bit byte-to-bit relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_8_B2B,
+
+  /* This is a 13 bit pcode relocation for the MCS-51/8051 microcontroller.  */
+  BFD_RELOC_I51_13_PCODE,
+
   /* This is a 16 bit reloc for the AVR that stores 8 bit pc relative
      short offset into 7 bits.  */
   BFD_RELOC_AVR_7_PCREL,
diff --git a/bfd/config.bfd b/bfd/config.bfd
index 4e5a838fe4..e50b2d6371 100644
--- a/bfd/config.bfd
+++ b/bfd/config.bfd
@@ -349,6 +349,10 @@ case "${targ}" in
     ;;
 #endif /* BFD64 */
 
+  i51-*-*)
+    targ_defvec=bfd_elf32_i51_vec
+    ;;
+
   am33_2.0-*-linux*)
     targ_defvec=am33_elf32_linux_vec
     ;;
diff --git a/bfd/configure b/bfd/configure
index 435b8e2071..49754952ec 100755
--- a/bfd/configure
+++ b/bfd/configure
@@ -15568,6 +15568,7 @@ do
     i386_pe_big_vec)		 tb="$tb pe-i386.lo peigen.lo $coff" ;;
     i386_pei_vec)		 tb="$tb pei-i386.lo peigen.lo $coff" ;;
     iamcu_elf32_vec)		 tb="$tb elf32-i386.lo $elfxx_x86 elf32.lo $elf" ;;
+    bfd_elf32_i51_vec)		 tb="$tb elf32-i51.lo elf32.lo $elf" ;;
     ia64_elf32_be_vec)		 tb="$tb elf32-ia64.lo elfxx-ia64.lo elf32.lo $elf" ;;
     ia64_elf32_hpux_be_vec)	 tb="$tb elf32-ia64.lo elfxx-ia64.lo elf32.lo $elf" ;;
     ia64_elf64_be_vec)		 tb="$tb elf64-ia64.lo elfxx-ia64.lo elf64.lo $elf"; target_size=64 ;;
diff --git a/bfd/configure.ac b/bfd/configure.ac
index 5879a04785..c79f691940 100644
--- a/bfd/configure.ac
+++ b/bfd/configure.ac
@@ -493,6 +493,7 @@ do
     i386_pe_big_vec)		 tb="$tb pe-i386.lo peigen.lo $coff" ;;
     i386_pei_vec)		 tb="$tb pei-i386.lo peigen.lo $coff" ;;
     iamcu_elf32_vec)		 tb="$tb elf32-i386.lo $elfxx_x86 elf32.lo $elf" ;;
+    bfd_elf32_i51_vec)		 tb="$tb elf32-i51.lo elf32.lo $elf" ;;
     ia64_elf32_be_vec)		 tb="$tb elf32-ia64.lo elfxx-ia64.lo elf32.lo $elf" ;;
     ia64_elf32_hpux_be_vec)	 tb="$tb elf32-ia64.lo elfxx-ia64.lo elf32.lo $elf" ;;
     ia64_elf64_be_vec)		 tb="$tb elf64-ia64.lo elfxx-ia64.lo elf64.lo $elf"; target_size=64 ;;
diff --git a/bfd/libbfd.h b/bfd/libbfd.h
index bab1f71f7e..4bd2372ecc 100644
--- a/bfd/libbfd.h
+++ b/bfd/libbfd.h
@@ -2383,6 +2383,17 @@ static const char *const bfd_reloc_code_real_names[] = { "@@uninitialized@@",
   "BFD_RELOC_MMIX_REG",
   "BFD_RELOC_MMIX_BASE_PLUS_OFFSET",
   "BFD_RELOC_MMIX_LOCAL",
+  "BFD_RELOC_I51_R1",
+  "BFD_RELOC_I51_R3",
+  "BFD_RELOC_I51_7_PCREL",
+  "BFD_RELOC_I51_11",
+  "BFD_RELOC_I51_8_BIT",
+  "BFD_RELOC_I51_8",
+  "BFD_RELOC_I51_8_LOW",
+  "BFD_RELOC_I51_8_HIGH",
+  "BFD_RELOC_I51_16",
+  "BFD_RELOC_I51_8_B2B",
+  "BFD_RELOC_I51_13_PCODE",
   "BFD_RELOC_AVR_7_PCREL",
   "BFD_RELOC_AVR_13_PCREL",
   "BFD_RELOC_AVR_16_PM",
diff --git a/bfd/reloc.c b/bfd/reloc.c
index c9d53bb9e1..283624bf20 100644
--- a/bfd/reloc.c
+++ b/bfd/reloc.c
@@ -4769,6 +4769,51 @@ ENUMDOC
   This relocation is an assertion that the expression is not allocated
   as a global register.  It does not modify contents.
 
+ENUM
+  BFD_RELOC_I51_R1
+ENUMDOC
+  This is a 1 bit relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_R3
+ENUMDOC
+  This is a 3 bit relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_7_PCREL
+ENUMDOC
+  This is a 7 bit PC relative relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_11
+ENUMDOC
+  This is an 11 bit relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_8_BIT
+ENUMDOC
+  This is an 8 bit relocation for bit addressing in the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_8
+ENUMDOC
+  This is an 8 bit relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_8_LOW
+ENUMDOC
+  This is an 8 bit low byte relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_8_HIGH
+ENUMDOC
+  This is an 8 bit high byte relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_16
+ENUMDOC
+  This is a 16 bit relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_8_B2B
+ENUMDOC
+  This is an 8 bit byte-to-bit relocation for the MCS-51/8051 microcontroller.
+ENUM
+  BFD_RELOC_I51_13_PCODE
+ENUMDOC
+  This is a 13 bit pcode relocation for the MCS-51/8051 microcontroller.
+
 ENUM
   BFD_RELOC_AVR_7_PCREL
 ENUMDOC
diff --git a/bfd/targets.c b/bfd/targets.c
index 23333701ec..a344f9ee87 100644
--- a/bfd/targets.c
+++ b/bfd/targets.c
@@ -757,6 +757,7 @@ extern const bfd_target i386_pe_vec;
 extern const bfd_target i386_pe_big_vec;
 extern const bfd_target i386_pei_vec;
 extern const bfd_target iamcu_elf32_vec;
+extern const bfd_target bfd_elf32_i51_vec;
 extern const bfd_target ia64_elf32_be_vec;
 extern const bfd_target ia64_elf32_hpux_be_vec;
 extern const bfd_target ia64_elf64_be_vec;
@@ -1105,6 +1106,7 @@ static const bfd_target * const _bfd_target_vector[] =
 	&i386_pei_vec,
 
 	&iamcu_elf32_vec,
+	&bfd_elf32_i51_vec,
 
 #ifdef BFD64
 	&bpf_elf64_be_vec,
diff --git a/config.sub b/config.sub
index 4aaae46f6f..41f155cb41 100755
--- a/config.sub
+++ b/config.sub
@@ -1317,6 +1317,7 @@ case $cpu-$vendor in
 			| hppa64 \
 			| i*86 \
 			| i370 \
+			| i51 \
 			| i860 \
 			| i960 \
 			| ia16 \
diff --git a/gas/configure.tgt b/gas/configure.tgt
index 7c81bffa16..417dc1685c 100644
--- a/gas/configure.tgt
+++ b/gas/configure.tgt
@@ -225,6 +225,8 @@ case ${generic_target} in
   h8300-*-elf)				fmt=elf ;;
   h8300-*-linux*)			fmt=elf em=linux ;;
 
+  i51-*-*)				fmt=elf ;;
+
   i386-*-beospe*)			fmt=coff em=pe ;;
   i386-*-beos*)				fmt=elf ;;
   i386-*-elfiamcu)			fmt=elf arch=iamcu ;;
diff --git a/ld/Makefile.am b/ld/Makefile.am
index ccff928992..c6ee2620be 100644
--- a/ld/Makefile.am
+++ b/ld/Makefile.am
@@ -284,6 +284,7 @@ ALL_EMULATION_SOURCES = \
 	eelf_i386_sol2.c \
 	eelf_i386_vxworks.c \
 	eelf_iamcu.c \
+	eelf32i51.c \
 	eelf_s390.c \
 	eh8300elf.c \
 	eh8300elf_linux.c \
diff --git a/ld/configure.tgt b/ld/configure.tgt
index ea01ccf9a1..a612f23771 100644
--- a/ld/configure.tgt
+++ b/ld/configure.tgt
@@ -337,6 +337,8 @@ h8300-*-linux*)
 			targ_emul=h8300elf_linux;
 			targ_extra_emuls="h8300helf_linux h8300self_linux h8300sxelf_linux"
 			;;
+i51-*-*)		targ_emul=elf32i51
+			;;
 hppa*64*-*-linux-*)	targ_emul=hppa64linux
 			;;
 hppa*64*-hpux*)		targ_emul=elf64hppa
diff --git a/opcodes/Makefile.am b/opcodes/Makefile.am
index ef4b0dfb01..29a9fb914b 100644
--- a/opcodes/Makefile.am
+++ b/opcodes/Makefile.am
@@ -156,6 +156,7 @@ TARGET32_LIBOPCODES_CFILES = \
 	h8300-dis.c \
 	hppa-dis.c \
 	i386-dis.c \
+	i51-dis.c \
 	ip2k-asm.c \
 	ip2k-desc.c \
 	ip2k-dis.c \
diff --git a/opcodes/configure b/opcodes/configure
index 210ec0741d..4b740f5cc7 100755
--- a/opcodes/configure
+++ b/opcodes/configure
@@ -13971,6 +13971,7 @@ if test x${all_targets} = xfalse ; then
 	bfd_hppa_arch)		ta="$ta hppa-dis.lo" ;;
 	bfd_i386_arch|bfd_iamcu_arch)
 				ta="$ta i386-dis.lo" ;;
+	bfd_i51_arch)		ta="$ta i51-dis.lo" ;;
 	bfd_ia64_arch)		ta="$ta ia64-dis.lo ia64-opc.lo" ;;
 	bfd_ip2k_arch)		ta="$ta ip2k-asm.lo ip2k-desc.lo ip2k-dis.lo ip2k-ibld.lo ip2k-opc.lo" using_cgen=yes ;;
 	bfd_epiphany_arch)	ta="$ta epiphany-asm.lo epiphany-desc.lo epiphany-dis.lo epiphany-ibld.lo epiphany-opc.lo" using_cgen=yes ;;
diff --git a/opcodes/configure.ac b/opcodes/configure.ac
index e31ddde9e5..9b80e815be 100644
--- a/opcodes/configure.ac
+++ b/opcodes/configure.ac
@@ -285,6 +285,7 @@ if test x${all_targets} = xfalse ; then
 	bfd_hppa_arch)		ta="$ta hppa-dis.lo" ;;
 	bfd_i386_arch|bfd_iamcu_arch)
 				ta="$ta i386-dis.lo" ;;
+	bfd_i51_arch)		ta="$ta i51-dis.lo" ;;
 	bfd_ia64_arch)		ta="$ta ia64-dis.lo ia64-opc.lo" ;;
 	bfd_ip2k_arch)		ta="$ta ip2k-asm.lo ip2k-desc.lo ip2k-dis.lo ip2k-ibld.lo ip2k-opc.lo" using_cgen=yes ;;
 	bfd_epiphany_arch)	ta="$ta epiphany-asm.lo epiphany-desc.lo epiphany-dis.lo epiphany-ibld.lo epiphany-opc.lo" using_cgen=yes ;;
diff --git a/ld/Makefile.in b/ld/Makefile.in
--- a/ld/Makefile.in
+++ b/ld/Makefile.in
@@ -797,6 +797,7 @@
 	eelf_i386_sol2.c \
 	eelf_i386_vxworks.c \
 	eelf_iamcu.c \
+	eelf32i51.c \
 	eelf_s390.c \
 	eh8300elf.c \
 	eh8300elf_linux.c \
